### PR TITLE
fix(storage): monitored resource detection

### PR DESCRIPTION
**Pull request by frankyn**
Summary of changes
- Fixed resource attribute names detected by GCP resource detector by using `_` instead of `.`, e.g. "host_id" instead of "host.id" resulting in storageClient always using default attribute values.
- Prioritize project ID from ADC before using resource detector project
- Refactored preparedResource into storageMonitoredResource to separate responsibilities
- Explicitly name gRPC metrics enabled instead of relying on `opentelemetry.DefaultMetrics()`  
- Introduced integration tests for gRPC Metrics are collected through a secondary Reader 
- Used struct embeddings to reduce boiler plate on log suppression exporter (thanks @BrennaEpp)

### CODE REVIEW

**Comment by frankyn** [comment]
Spot checked resulting monitored resource values to double check: 
image

File: storage/grpc_client.go [hunk 1]
```diff
@@ -118,6 +119,24 @@ type grpcStorageClient struct {
 	settings *settings
 }
 
+func enableClientMetrics(ctx context.Context, s *settings, config storageConfig) (*metricsContext, error) {
+	var project string
+	// TODO: use new auth client
```
Comment (tritone) [inline storage/grpc_client.go:124]: do we have a bug for this?

File: storage/grpc_metrics.go [hunk 2]
```diff
@@ -39,98 +39,76 @@ const (
...
-		}
+func (smr *storageMonitoredResource) exporter() (metric.Exporter, error) {
+	exporter, err := mexporter.New(
+		mexporter.WithProjectID(smr.project),
+		mexporter.WithMetricDescriptorTypeFormatter(metricFormatter),
+		mexporter.WithCreateServiceTimeSeries(),
+		mexporter.WithMonitoredResourceDescription(monitoredResourceName, []string{"project_id", "location", "cloud_platform", "host_id", "instance_id", "api"}),
+	)
+	if err != nil {
+		return nil, err
```
Comment (tritone) [inline storage/grpc_metrics.go:61]: Let's wrap this with a string like "storage: creating metrics exporter: %w"

File: storage/grpc_metrics.go [hunk 3]
```diff
@@ -39,98 +39,76 @@ const (
...
+	smr := &storageMonitoredResource{
+		instance: uuid.New().String(),
+		api:      api,
+		project:  project,
+	}
 	s := detectedAttrs.Set()
-	p, present := s.Value("cloud.account.id")
-	if present {
-		preparedResource.projectToUse = p.AsString()
+	if p, present := s.Value("cloud.account.id"); present && smr.project == "" {
```
Comment (tritone) [inline storage/grpc_metrics.go:77]: Maybe add a comment about what this is trying to do?

File: storage/grpc_metrics_test.go [hunk 4]
```diff
@@ -34,15 +34,19 @@ func TestMetricFormatter(t *testing.T) {
...
 		desc               string
+		project            string
+		api                string
 		detectedAttributes []attribute.KeyValue
 		wantAttributes     attribute.Set
 	}{
 		{
-			desc: "default values set when GCP attributes are not detected",
+			desc:    "default values set when GCP attributes are not detected",
+			project: "project-id",
```
Comment (tritone) [inline storage/grpc_metrics_test.go:48]: project and api seem the same in all test cases, should they just be removed as fields here? Or did you want to test other values?

File: storage/integration_test.go [hunk 5]
```diff
@@ -413,6 +415,114 @@ func TestIntegration_DoNotDetectDirectConnectivityWhenDisabled(t *testing.T) {
...
+		}
+		for k, v := range metricCheck {
+			if !v {
+				t.Errorf("metric %v not found", k)
+			}
+		}
+	}, withTestMetricReader(mr))
+}
+
+func TestIntegration_MetricsEnablementInGCE(t *testing.T) {
```
Comment (tritone) [inline storage/integration_test.go:451]: Do we expect this test to run without skipping in Kokoro?

File: storage/integration_test.go [hunk 6]
```diff
@@ -413,6 +415,114 @@ func TestIntegration_DoNotDetectDirectConnectivityWhenDisabled(t *testing.T) {
...
+		metricCheck := map[string]bool{
+			"grpc.client.attempt.started":                            false,
+			"grpc.client.attempt.duration":                           false,
+			"grpc.client.attempt.sent_total_compressed_message_size": false,
+			"grpc.client.attempt.rcvd_total_compressed_message_size": false,
+			"grpc.client.call.duration":                              false,
+			"grpc.lb.rls.cache_entries":                              false,
+			"grpc.lb.rls.cache_size":                                 false,
+			"grpc.lb.rls.default_target_picks":                       false,
+			// TODO: determine a way to force these metrics to be collected
```
Comment (tritone) [inline storage/integration_test.go:505]: Remove commented values here; please file an issue if there is other stuff we need to add later.

**Review by tritone — COMMENTED** [review]
Few minor questions/nits, otherwise looks good.

File: storage/integration_test.go (hunk 5 above)
Comment (tritone) [reply storage/integration_test.go:454]: No, looks fine to me, just wanted to confirm.

**Review by tritone — APPROVED** [review]

**Pull request merged**